### PR TITLE
feat(dns): make `{tokio,async_std}::Transport::custom` infallible

### LINKED ISSUE #4462: dns: make `DnsConfig::custom` infallible
With https://github.com/libp2p/rust-libp2p/pull/4441/ we can now make the method infallible.

### PR DESCRIPTION
**Pull request by p0mvn**
## Description

Resolves: #4462.

## Notes & open questions

N/A

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: transports/dns/CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,9 @@
 
 [PR 3715]: https://github.com/libp2p/rust-libp2p/pull/3715
 
+- make DnsConfig::custom infallible
+[PR 4464]: https://github.com/libp2p/rust-libp2p/pull/4464
```
Comment (thomaseizinger) [inline transports/dns/CHANGELOG.md]: This is a breaking change so we need this to be in a `0.41.0` section and the version in the `Cargo.toml` needs to be bumped too! :)

See https://github.com/libp2p/rust-libp2p/blob/master/CONTRIBUTING.md#write-changelog-entries-for-user-facing-changes and the other linked docs.

File: transports/dns/CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,9 @@
 
 [PR 3715]: https://github.com/libp2p/rust-libp2p/pull/3715
 
+- make DnsConfig::custom infallible
```
Comment (thomaseizinger) [inline transports/dns/CHANGELOG.md]:
```suggestion
- Make `DnsConfig::custom` constructor infallible.
  See [PR 4464].
```

**Review by thomaseizinger — COMMENTED** [review]
Thanks!

I am pretty sure this method is used in various places, can you adjust those too? I'd recommend using `cargo clippy --all-features --all-targets` to check.

**Comment by thomaseizinger** [comment]
This is a breaking change so I am queuing it for the next milestone. FYI: We mark those as draft to not merge them accidentally!

**Comment by mxinden** [comment]
Meta: very much appreciate the help here!

File: transports/dns/CHANGELOG.md [hunk 3]
```diff
@@ -1,13 +1,15 @@
+## 0.41.0
```
Comment (thomaseizinger) [inline transports/dns/CHANGELOG.md]:
```suggestion
## 0.41.0 - unreleased
```

File: transports/dns/CHANGELOG.md [hunk 4]
```diff
@@ -1,13 +1,15 @@
...
+
+- Make DnsConfig::custom constructor infallible
+[PR 4464]: https://github.com/libp2p/rust-libp2p/pull/4464
```
Comment (thomaseizinger) [inline transports/dns/CHANGELOG.md]:
```suggestion
- Make DnsConfig::custom constructor infallible.
  See [PR 4464].

[PR 4464]: https://github.com/libp2p/rust-libp2p/pull/4464
```

**Review by thomaseizinger — DISMISSED** [review]

**Review by thomaseizinger — APPROVED** [review]

**Pull request merged**